### PR TITLE
Add option to turn off optimization for X86 assembler

### PR DESCRIPTION
**Pull request by kongy**
There are use cases that we are not expecting the assembler to produce the exact instructions without any optimizations.

### CODE REVIEW

File: llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp [hunk 1]
```diff
@@ -48,6 +48,10 @@ static cl::opt<bool> LVIInlineAsmHardening(
     cl::desc("Harden inline assembly code that may be vulnerable to Load Value"
              " Injection (LVI). This feature is experimental."), cl::Hidden);
 
+static cl::opt<bool> AsmOptimize("x86-inline-asm-optimize", cl::init(true),
```
Comment (KanRobert) [inline llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:51]: It's not inline asm opt. It works on assembly.

Comment (KanRobert) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:51]: clang format

**Comment by KanRobert** [comment]
I think it should be a target-independent flag. Maybe we should add a flag "O0" for llvm-mc ?

File: llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp (hunk 1 above)
Comment (kongy) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp]: Done.

**Comment by kongy** [comment]
> I think it should be a target-independent flag. Maybe we should add a flag "O0" for llvm-mc ?

This is a great suggestion, but can we add this as a hidden flag for now to unblock Android's integration?

File: llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp [hunk 2]
```diff
@@ -48,6 +48,10 @@ static cl::opt<bool> LVIInlineAsmHardening(
     cl::desc("Harden inline assembly code that may be vulnerable to Load Value"
              " Injection (LVI). This feature is experimental."), cl::Hidden);
 
+static cl::opt<bool> AsmOptimize("x86-asm-optimize", cl::init(true),
+                                 cl::desc("Optimize X86 inline assembly code."),
```
Comment (KanRobert) [inline llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:52]: Comment needs to be updated too.

Comment (kongy) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp]: Done.

**Comment by phoebewang** [comment]
The intention is not quite clear to me. For example, if you just want to always get VEX3 encoding, you can use `{vex3}` before the instruction. It is officially supported by both llvm-mc and gas. It looks less optimization than assembler preference to me.

**Comment by kongy** [comment]
> The intention is not quite clear to me. For example, if you just want to always get VEX3 encoding, you can use `{vex3}` before the instruction. It is officially supported by both llvm-mc and gas. It looks less optimization than assembler preference to me.

We have tests that use clang assembler as a verification tool, to verify that our assembler generates the correct (i.e. the same) output. Therefore, clang assembler performing optimisations is undesired.

File: llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp [hunk 3]
```diff
@@ -3670,11 +3674,11 @@ bool X86AsmParser::ParseInstruction(ParseInstructionInfo &Info, StringRef Name,
 }
 
 bool X86AsmParser::processInstruction(MCInst &Inst, const OperandVector &Ops) {
-  if (ForcedVEXEncoding != VEXEncoding_VEX3 &&
+  if (AsmOptimize && ForcedVEXEncoding != VEXEncoding_VEX3 &&
       X86::optimizeInstFromVEX3ToVEX2(Inst, MII.get(Inst.getOpcode())))
     return true;
 
-  if (X86::optimizeShiftRotateWithImmediateOne(Inst))
+  if (AsmOptimize && X86::optimizeShiftRotateWithImmediateOne(Inst))
```
Comment (topperc) [inline llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3681]: @kongy does your assembler not optimize shifts/rotates by 1 to use the short form? That seems like a pretty basic optimization that has been around for x86 for a very long time.

Comment (kongy) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3681]: After some investigation, our assembler (Android ART) does implement the optimizeShiftRotateWithImmediateOne optimization. I guess we only need to turn off the VEX3ToVEX2 optimization.

Comment (KanRobert) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3681]: Do you still need this patch? If so, I guess you might need to change it to "x86-asm-vex3-to-vex2"?

Comment (kongy) [reply llvm/lib/Target/X86/AsmParser/X86AsmParser.cpp:3681]: We ended up implementing the same optimization in our assembler. This change is no longer required.

**Pull request closed without merging**